src/promptflow-tools/promptflow/tools/common.py

```diff
@@ -56,15 +61,47 @@ def __str__(self):
         return "\n".join(map(str, self))
 
 
-def validate_role(role: str, valid_roles: List[str] = None):
+class PromptResult(str):
+    def __init__(self, string):
+        super().__init__()
+        self.original_string = string
+        self.escaped_string = ""
+        self.escaped_mapping = {}
+
+    def get_escape_string(self) -> str:
+        return self.escaped_string
+
+    def set_escape_string(self, escaped_string: str):
+        self.escaped_string = escaped_string
+
+    def get_escape_mapping(self) -> dict:
+        return self.escaped_mapping
+
+    def set_escape_mapping(self, escape_mapping: dict):
+        self.escaped_mapping = escape_mapping
+
+    def need_to_escape(self) -> bool:
+        return bool(self.escaped_mapping)
+
+    def merge_escape_mapping_of_prompt_results(self, **kwargs):
+        prompt_result_escape_dict = Escaper.merge_escape_mapping_of_prompt_results(**kwargs)
+        self.escaped_mapping.update(prompt_result_escape_dict)
+
+    def merge_escape_mapping_of_flow_inputs(self, inputs_to_escape: list, **kwargs):
+        flow_inputs_escape_dict = Escaper.build_flow_inputs_escape_dict(inputs_to_escape=inputs_to_escape, **kwargs)
+        self.escaped_mapping.update(flow_inputs_escape_dict)
+
+
+def validate_role(role: str, valid_roles: List[str] = None, escape_dict: dict = {}):
     if not valid_roles:
         valid_roles = VALID_ROLES
 
     if role not in valid_roles:
         valid_roles_str = ','.join([f'\'{role}:\\n\'' for role in valid_roles])
+        unescaped_invalid_role = Escaper._unescape_roles(role, escape_dict)
```

Comment:
add comment here to explain why must unescape here.

Reply:
added